src/main/scala/com/mongodb/spark/rdd/partitioner/BsonValueOrdering.scala

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2016 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.spark.rdd.partitioner
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+
+import org.bson._
+
+trait BsonValueOrdering extends Ordering[BsonValue] {
+
+  // scalastyle:off cyclomatic.complexity
+  private val bsonTypeComparisonMap: Map[BsonType, Int] = Map(
+    BsonType.MIN_KEY -> 1,
+    BsonType.NULL -> 2,
+    BsonType.INT32 -> 3,
+    BsonType.INT64 -> 3,
+    BsonType.DOUBLE -> 3,
+    BsonType.SYMBOL -> 4,
+    BsonType.STRING -> 4,
+    BsonType.DOCUMENT -> 5,
+    BsonType.ARRAY -> 6,
+    BsonType.BINARY -> 7,
+    BsonType.OBJECT_ID -> 8,
+    BsonType.BOOLEAN -> 9,
+    BsonType.DATE_TIME -> 10,
+    BsonType.TIMESTAMP -> 11,
+    BsonType.REGULAR_EXPRESSION -> 12,
+    BsonType.MAX_KEY -> 13
+  )
+
+  /**
+   *
+   * Returns an integer whose sign communicates how x compares to y.
+   *
+   * The result sign has the following meaning:
+   *
+   * - negative if x < y
+   * - positive if x > y
+   * - zero otherwise (if x == y)
+   *
+   * When comparing values of different BSON types, MongoDB uses the following comparison order, from lowest to highest:
+   *
+   * 1. MinKey (internal type)
+   * 2. Null
+   * 3. Numbers (ints, longs, doubles)
+   * 4. Symbol, String
+   * 5. Object
+   * 6. Array
+   * 7. BinData
+   * 8. ObjectId
+   * 9. Boolean
+   * 10. Date
+   * 11. Timestamp
+   * 12. Regular Expression
+   * 13. MaxKey (internal type)
+   */
+  override def compare(x: BsonValue, y: BsonValue): Int = {
+    (x.getBsonType, y.getBsonType) match {
+      case (BsonType.MIN_KEY, BsonType.MIN_KEY)     => 0
+      case (BsonType.NULL, BsonType.NULL)           => 0
+      case (isBsonNumber(), isBsonNumber())         => x.asNumber.doubleValue.compareTo(y.asNumber.doubleValue)
+      case (BsonType.STRING, BsonType.STRING)       => x.asString.getValue.compareTo(y.asString.getValue)
+      case (BsonType.SYMBOL, BsonType.STRING)       => x.asSymbol.getSymbol.compareTo(y.asString.getValue)
```

Comment:
To be consistent with the server, I think this should do a straight byte comparison of the UTF-8 encoded values of each string, rather than of the String values themselves.